File: swim-js/swim-core-js/@swim/streamlet/main/AbstractInlet.ts
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Inlet} from "./Inlet";
import {Outlet} from "./Outlet";

export abstract class AbstractInlet<I> implements Inlet<I> {
  /** @hidden */
  protected _input: Outlet<I> | null;
  /** @hidden */
  protected _version: number;

  constructor() {
    this._input = null;
    this._version = -1;
  }

  input(): Outlet<I> | null {
    return this._input;
  }

  bindInput(input: Outlet<I> | null): void {
    if (this._input !== null) {
      this._input.unbindOutput(this);
    }
    this._input = input;
    if (this._input !== null) {
      this._input.bindOutput(this);
    }
  }

  unbindInput(): void {
    if (this._input !== null) {
      this._input.unbindOutput(this);
    }
    this._input = null;
  }

  disconnectInputs(): void {
    const input = this._input;
    if (input !== null) {
      input.unbindOutput(this);
      this._input = null;
      input.disconnectInputs();
    }
  }

  disconnectOutputs(): void {
    // nop
  }

  invalidateOutput(): void {
    if (this._version >= 0) {
      this.willInvalidateOutput();
      this._version = -1;
      this.onInvalidateOutput();
      this.didInvalidateOutput();
    }
  }

  reconcileOutput(version: number): void {
    if (this._version < 0) {
      this.willReconcileOutput(version);
      this._version = version;
      if (this._input !== null) {
        this._input.reconcileInput(version);
      }
      this.onReconcileOutput(version);
      this.didReconcileOutput(version);
    }
  }

  protected willInvalidateOutput(): void {
    // stub
  }

  protected onInvalidateOutput(): void {
    // stub
  }

  protected didInvalidateOutput(): void {
    // stub
  }

  protected willReconcileOutput(version: number): void {
    // stub
  }

  protected onReconcileOutput(version: number): void {
    // stub
  }

  protected didReconcileOutput(version: number): void {
    // stub
  }
}

File: swim-system-java/swim-core-java/swim.streamlet/src/main/java/swim/streamlet/AbstractInlet.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.streamlet;

public abstract class AbstractInlet<I> implements Inlet<I> {
  protected Outlet<? extends I> input;
  protected int version;

  public AbstractInlet() {
    this.input = null;
    this.version = -1;
  }

  @Override
  public Outlet<? extends I> input() {
    return this.input;
  }

  @Override
  public void bindInput(Outlet<? extends I> input) {
    if (this.input != null) {
      this.input.unbindOutput(this);
    }
    this.input = input;
    if (this.input != null) {
      this.input.bindOutput(this);
    }
  }

  @Override
  public void unbindInput() {
    if (this.input != null) {
      this.input.unbindOutput(this);
    }
    this.input = null;
  }

  @Override
  public void disconnectInputs() {
    final Outlet<? extends I> input = this.input;
    if (input != null) {
      input.unbindOutput(this);
      this.input = null;
      input.disconnectInputs();
    }
  }

  @Override
  public void disconnectOutputs() {
    // nop
  }

  @Override
  public void invalidateOutput() {
    if (this.version >= 0) {
      willInvalidateOutput();
      this.version = -1;
      onInvalidateOutput();
      didInvalidateOutput();
    }
  }

  @Override
  public void reconcileOutput(int version) {
    if (this.version < 0) {
      willReconcileOutput(version);
      this.version = version;
      if (this.input != null) {
        this.input.reconcileInput(version);
      }
      onReconcileOutput(version);
      didReconcileOutput(version);
    }
  }

  protected void willInvalidateOutput() {
    // stub
  }

  protected void onInvalidateOutput() {
    // stub
  }

  protected void didInvalidateOutput() {
    // stub
  }

  protected void willReconcileOutput(int version) {
    // stub
  }

  protected void onReconcileOutput(int version) {
    // stub
  }

  protected void didReconcileOutput(int version) {
    // stub
  }
}
